timesource/timesource.go

```diff
@@ -0,0 +1,158 @@
+package timesource
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/beevik/ntp"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	// DefaultServer will be internally resolved to the closest available.
+	// also it rarely queries same server more than once.
+	DefaultServer = "pool.ntp.org"
+
+	// DefaultAttempts defines how many servers we will query
+	DefaultAttempts = 3
+
+	// DefaultUpdatePeriod defines how often time will be queried from ntp.
+	DefaultUpdatePeriod = 2 * time.Minute
+)
+
+type ntpQuery func(string) (*ntp.Response, error)
+
+type queryResponse struct {
+	Offset time.Duration
+	Error  error
+}
+
+type multiRPCError []error
+
+func (e multiRPCError) Error() string {
+	var b bytes.Buffer
+	b.WriteString("RPC failed: ")
+	more := false
+	for _, err := range e {
+		if more {
+			b.WriteString("; ")
+		}
+		b.WriteString(err.Error())
+		more = true
+	}
+	b.WriteString(".")
+	return b.String()
+}
+
+func computeOffset(timeQuery ntpQuery, server string, attempts int) (time.Duration, error) {
+	responses := make(chan queryResponse, attempts)
+	for i := 0; i < attempts; i++ {
+		go func() {
+			// ntp.Query default timeout is 5s
+			response, err := timeQuery(server)
+			if err != nil {
+				responses <- queryResponse{Error: err}
+				return
+			}
+			responses <- queryResponse{Offset: response.ClockOffset}
+		}()
+	}
+	var (
+		rpcErrors multiRPCError
+		offsets   []time.Duration
+		collected int
+	)
+	for response := range responses {
+		if response.Error != nil {
+			rpcErrors = append(rpcErrors, response.Error)
+		} else {
+			offsets = append(offsets, response.Offset)
+		}
+		collected++
+		if collected == attempts {
+			break
+		}
+	}
+	if len(rpcErrors) != 0 {
+		return 0, rpcErrors
+	}
+	sort.SliceStable(offsets, func(i, j int) bool {
+		return offsets[i] > offsets[j]
+	})
+	return offsets[attempts/2], nil
```

Comment:
If it's a median, for the slice of even length, it should calculate an average of middle elements.

Reply:
why?

Reply:
It's just a median definition. In case of four attempts and results like `[4 5 9 11]`, `9` is not really the middle value. `5+9/2` is more accurate.